File: native/macos/Sources/MightyCore/CLIUpdateService.swift
import Foundation

public struct CLIUpdateInstallation: Sendable, Equatable {
    public var provider: String
    public var version: String?
    /// native, brew-cask, brew-formula, npm, unknown, or missing.
    public var method: String
    public var executablePath: String?
    public var canUpdate: Bool
    public var detail: String

    public init(provider: String, version: String? = nil, method: String = "unknown", executablePath: String? = nil, canUpdate: Bool = false, detail: String) {
        self.provider = provider; self.version = version; self.method = method
        self.executablePath = executablePath; self.canUpdate = canUpdate; self.detail = detail
    }
}

public struct CLIUpdateResult: Sendable, Equatable {
    public var provider: String
    /// updated, current, skipped, failed, cancelled, or busy.
    public var status: String
    public var beforeVersion: String?
    public var afterVersion: String?
    public var method: String
    public var detail: String
    /// Bounded diagnostic output. Present only on an explicit details action.
    public var output: String

    public init(provider: String, status: String, beforeVersion: String? = nil, afterVersion: String? = nil, method: String = "unknown", detail: String, output: String = "") {
        self.provider = provider; self.status = status; self.beforeVersion = beforeVersion
        self.afterVersion = afterVersion; self.method = method; self.detail = detail; self.output = output
    }
}

private struct CLIUpdateConfiguration: Sendable {
    let environment: [String: String]
    let home: URL
    let metadataTimeout: TimeInterval
    let updateTimeout: TimeInterval
}

private struct CLIUpdateInvocation: Sendable {
    let executable: URL
    let arguments: [String]
    var environment: [String: String]
}

private struct CLIUpdatePlan: Sendable {
    let installation: CLIUpdateInstallation
    let invocation: CLIUpdateInvocation?
}

/// Updates only an already installed, recognized CLI through its own installer.
/// No shell command construction, sudo, fresh installations, or config writes.
public actor CLIUpdateService {
    private let configuration: CLIUpdateConfiguration
    private var active: (id: UUID, task: Task<CLIUpdateResult, Never>)?
    private var closing = false

    public init(environment: [String: String]? = nil) {
        configuration = CLIUpdateConfiguration(environment: environment ?? ProviderService.runtimeEnvironment(), home: FileManager.default.homeDirectoryForCurrentUser, metadataTimeout: 4, updateTimeout: 300)
    }

    // Filesystem fixtures use the real process implementation with private roots
    // and short deadlines. These overrides are not exposed to the app or wire.
    // The probe only has to beat a hung CLI. Two seconds was not enough for a
    // shell fixture when the whole suite spawns processes in parallel, and a
    // missed probe reads as "unknown" rather than as a timeout.
    init(environment: [String: String], homeDirectory: URL, metadataTimeout: TimeInterval = 15, updateTimeout: TimeInterval = 30) {
        configuration = CLIUpdateConfiguration(environment: environment, home: homeDirectory, metadataTimeout: metadataTimeout, updateTimeout: updateTimeout)
    }

    public func inspect(provider: String) async -> CLIUpdateInstallation {
        guard !closing else { return CLIUpdateInstallation(provider: provider, detail: "앱이 종료 중입니다.") }
        do { return try await Self.plan(provider: provider, configuration: configuration).installation }
        catch { return CLIUpdateInstallation(provider: provider, detail: Task.isCancelled ? "설치 확인을 취소했습니다." : "CLI 설치 정보를 확인하지 못했습니다.") }
    }

    public func update(provider: String) async -> CLIUpdateResult {
        guard !closing else { return CLIUpdateResult(provider: provider, status: "cancelled", detail: "앱이 종료 중입니다.") }
        guard active == nil else { return CLIUpdateResult(provider: provider, status: "busy", detail: "다른 CLI를 업데이트하고 있습니다.") }
        guard !Task.isCancelled else { return CLIUpdateResult(provider: provider, status: "cancelled", detail: "업데이트를 취소했습니다.") }
        let id = UUID(), configuration = configuration
        let task = Task { await Self.performUpdate(provider: provider, configuration: configuration) }
        active = (id, task)
        let result = await withTaskCancellationHandler(operation: { await task.value }, onCancel: { task.cancel() })
        if active?.id == id { active = nil }
        return result
    }

    public func cancel() async {
        guard let operation = active else { return }
        operation.task.cancel(); _ = await operation.task.value
        if active?.id == operation.id { active = nil }
    }

    public func shutdown() async { closing = true; await cancel() }

    private static func performUpdate(provider: String, configuration: CLIUpdateConfiguration) async -> CLIUpdateResult {
        var installation = CLIUpdateInstallation(provider: provider, detail: "")
        do {
            try Task.checkCancellation()
            let plan = try await plan(provider: provider, configuration: configuration)
            installation = plan.installation
            try Task.checkCancellation()
            guard let invocation = plan.invocation else {
                return CLIUpdateResult(provider: provider, status: "skipped", beforeVersion: installation.version, method: installation.method, detail: installation.detail)
            }
            let result = try await ProcessCapture.run(executable: invocation.executable, arguments: invocation.arguments,
                environment: invocation.environment, timeout: configuration.updateTimeout, maximumBytes: 1024 * 1024)
            try Task.checkCancellation()
            let output = boundedOutput(result)
            guard result.exitCode == 0 else {
                return CLIUpdateResult(provider: provider, status: "failed", beforeVersion: installation.version, method: installation.method,
                    detail: "업데이트 명령이 종료 코드 \(result.exitCode)로 실패했습니다. 설치 권한이나 네트워크 상태를 확인하세요.", output: output)
            }
            let after = try await installedCommand(provider: provider, configuration: configuration)
            try Task.checkCancellation()
            guard let version = after?.version else {
                return CLIUpdateResult(provider: provider, status: "failed", beforeVersion: installation.version, method: installation.method,
                    detail: "업데이트 명령은 끝났지만 CLI 버전을 다시 확인하지 못했습니다.", output: output)
            }
            let changed = installation.version != version
            return CLIUpdateResult(provider: provider, status: changed ? "updated" : "current", beforeVersion: installation.version, afterVersion: version,
                method: installation.method, detail: changed ? "CLI를 업데이트했습니다." : "업데이트 명령을 완료했습니다. 설치된 버전은 동일합니다.", output: output)
        } catch {
            let cancelled = error is CancellationError || Task.isCancelled
            return CLIUpdateResult(provider: provider, status: cancelled ? "cancelled" : "failed", beforeVersion: installation.version, method: installation.method,
                detail: cancelled ? "업데이트를 취소했습니다." : String(error.localizedDescription.prefix(600)))
        }
    }

    private static func plan(provider: String, configuration: CLIUpdateConfiguration) async throws -> CLIUpdatePlan {
        guard ProviderOptions.ids.contains(provider) else {
            return CLIUpdatePlan(installation: CLIUpdateInstallation(provider: provider, detail: "지원하지 않는 CLI입니다."), invocation: nil)
        }
        guard let command = try await installedCommand(provider: provider, configuration: configuration) else {
            let present = executableCandidates(provider, environment: configuration.environment).contains { FileManager.default.isExecutableFile(atPath: $0.path) }
            return CLIUpdatePlan(installation: CLIUpdateInstallation(provider: provider, method: present ? "unknown" : "missing",
                detail: present ? "CLI 버전을 확인하지 못해 업데이트하지 않았습니다." : "설치된 CLI가 없어 건너뜁니다. 새로 설치하지 않습니다."), invocation: nil)
        }
        let resolved = command.executable.resolvingSymlinksInPath().standardizedFileURL
        var installation = CLIUpdateInstallation(provider: provider, version: command.version, executablePath: command.executable.path,
            detail: "수동 설치 또는 확인할 수 없는 설치 방식입니다. 기존 설치 방법으로 직접 업데이트하세요.")
        var environment = configuration.environment
        if let brew = brewInstallation(resolved, provider: provider) {
            installation.method = brew.cask ? "brew-cask" : "brew-formula"
            guard FileManager.default.isExecutableFile(atPath: brew.executable.path) else {
                installation.detail = "Homebrew 설치이지만 해당 Homebrew 실행 파일을 찾지 못했습니다."
                return CLIUpdatePlan(installation: installation, invocation: nil)
            }
            // Refreshing package metadata is allowed; do not clean up or upgrade
            // unrelated installed dependents as a side effect of this request.
            environment["HOMEBREW_NO_INSTALL_CLEANUP"] = "1"
            environment["HOMEBREW_NO_INSTALLED_DEPENDENTS_CHECK"] = "1"
            environment["HOMEBREW_NO_ANALYTICS"] = "1"
            environment["HOMEBREW_NO_ASK"] = "1"
            environment["NONINTERACTIVE"] = "1"
            installation.canUpdate = true; installation.detail = "설치된 Homebrew의 해당 \(brew.cask ? "cask" : "formula")만 업데이트합니다."
            return CLIUpdatePlan(installation: installation, invocation: CLIUpdateInvocation(executable: brew.executable,
                arguments: ["upgrade", brew.cask ? "--cask" : "--formula", brew.package], environment: environment))
        }
        if provider == "claude" {
            var nativeRoots = [configuration.home.appendingPathComponent(".local/share/claude/versions")]
            if let dataHome = configuration.environment["XDG_DATA_HOME"], validAbsolutePath(dataHome) {
                nativeRoots.append(URL(fileURLWithPath: dataHome).appendingPathComponent("claude/versions"))
            }
            if nativeRoots.contains(where: { resolved.deletingLastPathComponent().path == $0.resolvingSymlinksInPath().standardizedFileURL.path }) {
                installation.method = "native"; installation.canUpdate = true
                installation.detail = "Claude Code의 기본 업데이트 명령을 사용합니다."
                return CLIUpdatePlan(installation: installation, invocation: CLIUpdateInvocation(executable: command.executable, arguments: ["update"], environment: environment))
            }
        }
        if let package = npmInstallation(resolved, provider: provider) {
            installation.method = "npm"
            guard package.version.range(of: "\\A[0-9]+\\.[0-9]+\\.[0-9]+(?:\\+[0-9A-Za-z.-]+)?\\z", options: .regularExpression) != nil else {
                installation.detail = "시험판 또는 확인할 수 없는 npm 채널은 자동 변경하지 않습니다. 기존 채널에서 직접 업데이트하세요."
                return CLIUpdatePlan(installation: installation, invocation: nil)
            }
            guard let npm = npmRuntime(prefix: package.prefix, environment: environment) else {
                installation.detail = "npm 설치는 확인했지만 해당 설치를 업데이트할 Node.js/npm을 찾지 못했습니다."
                return CLIUpdatePlan(installation: installation, invocation: nil)
            }
            // npm lifecycle scripts and the CLI's env-node launcher use the same
            // Node installation as the npm process, including version managers.
            environment["PATH"] = npm.node.deletingLastPathComponent().path + ":" + (environment["PATH"] ?? "")
            installation.canUpdate = true; installation.detail = "기존 npm 설치 위치에서 공식 패키지만 업데이트합니다."
            return CLIUpdatePlan(installation: installation, invocation: CLIUpdateInvocation(executable: npm.node,
                arguments: [npm.script.path, "install", "--global", "--prefix", package.prefix.path, package.name + "@latest", "--no-audit", "--no-fund"], environment: environment))
        }
        return CLIUpdatePlan(installation: installation, invocation: nil)
    }

    private static func installedCommand(provider: String, configuration: CLIUpdateConfiguration) async throws -> ProviderCommand? {
        guard ProviderOptions.ids.contains(provider) else { return nil }
        for path in executableCandidates(provider, environment: configuration.environment) where FileManager.default.isExecutableFile(atPath: path.path) {
            try Task.checkCancellation()
            do {
                let result = try await ProcessCapture.run(executable: path, arguments: ["--version"], environment: configuration.environment,
                    timeout: configuration.metadataTimeout, maximumBytes: 16_384)
                guard result.exitCode == 0 else { continue }
                let text = String(decoding: result.stdout, as: UTF8.self).trimmingCharacters(in: .whitespacesAndNewlines)
                if !text.isEmpty { return ProviderCommand(provider: provider, executable: path, version: String(text.prefix(160))) }
            } catch { if error is CancellationError || Task.isCancelled { throw CancellationError() } }
        }
        return nil
    }

    private static func executableCandidates(_ name: String, environment: [String: String]) -> [URL] {
        var seen = Set<String>()
        return (environment["PATH"] ?? "").split(separator: ":").prefix(64).compactMap { entry in
            let path = String(entry)
            guard validAbsolutePath(path), seen.insert(path).inserted else { return nil }
            return URL(fileURLWithPath: path).appendingPathComponent(name)
        }
    }

    private static func validAbsolutePath(_ path: String) -> Bool { path.hasPrefix("/") && !path.contains("\0") }

    private static func brewInstallation(_ path: URL, provider: String) -> (executable: URL, package: String, cask: Bool)? {
        let package = ["claude": "claude-code", "codex": "codex", "gemini": "gemini-cli"][provider]!
        let parts = path.pathComponents
        guard let index = parts.firstIndex(where: { $0 == "Caskroom" || $0 == "Cellar" }), index > 1,
              parts.count > index + 3, parts[index + 1] == package else { return nil }
        let prefix = URL(fileURLWithPath: NSString.path(withComponents: Array(parts.prefix(index))))
        return (prefix.appendingPathComponent("bin/brew"), package, parts[index] == "Caskroom")
    }

    private static func npmInstallation(_ path: URL, provider: String) -> (prefix: URL, name: String, version: String)? {
        let name = ["claude": "@anthropic-ai/claude-code", "codex": "@openai/codex", "gemini": "@google/gemini-cli"][provider]!
        let components = path.pathComponents
        for index in components.indices where components[index] == "lib" && index + 3 < components.count && components[index + 1] == "node_modules" {
            let prefix = URL(fileURLWithPath: NSString.path(withComponents: Array(components.prefix(index))))
            let root = prefix.appendingPathComponent("lib/node_modules").appendingPathComponent(name)
            if officialPackage(root, name: name, binary: provider, executable: path) {
                return (prefix, name, packageMetadata(root)?["version"] as? String ?? "")
            }
        }
        return nil
    }

    private static func officialPackage(_ root: URL, name: String, binary: String, executable: URL) -> Bool {
        guard let json = packageMetadata(root), json["name"] as? String == name,
              let bin = (json["bin"] as? [String: String])?[binary] ?? json["bin"] as? String,
              !bin.hasPrefix("/"), !bin.contains("\0") else { return false }
        let target = root.appendingPathComponent(bin).resolvingSymlinksInPath().standardizedFileURL
        let packageRoot = root.resolvingSymlinksInPath().standardizedFileURL.path + "/"
        return target.path.hasPrefix(packageRoot) && target == executable
    }

    private static func packageMetadata(_ root: URL) -> [String: Any]? {
        let file = root.appendingPathComponent("package.json")
        guard let size = try? file.resourceValues(forKeys: [.fileSizeKey]).fileSize, size <= 262_144,
              let data = try? Data(contentsOf: file), data.count <= 262_144 else { return nil }
        return try? JSONSerialization.jsonObject(with: data) as? [String: Any]
    }

    private static func npmRuntime(prefix: URL, environment: [String: String]) -> (node: URL, script: URL)? {
        var candidates = [prefix.appendingPathComponent("bin/npm")]
        candidates += executableCandidates("npm", environment: environment)
        for launcher in candidates where FileManager.default.isExecutableFile(atPath: launcher.path) {
            let script = launcher.resolvingSymlinksInPath().standardizedFileURL
            var root = script.deletingLastPathComponent()
            for _ in 0..<7 {
                if officialPackage(root, name: "npm", binary: "npm", executable: script) {
                    var nodes = [launcher.deletingLastPathComponent().appendingPathComponent("node")]
                    nodes += executableCandidates("node", environment: environment)
                    if let node = nodes.first(where: { FileManager.default.isExecutableFile(atPath: $0.path) }) { return (node, script) }
                }
                let parent = root.deletingLastPathComponent(); if parent == root { break }; root = parent
            }
        }
        return nil
    }

    private static func boundedOutput(_ result: ProcessResult) -> String {
        let text = String(decoding: result.stdout, as: UTF8.self) + (result.stderr.isEmpty ? "" : "\n" + String(decoding: result.stderr, as: UTF8.self))
        let safe = String(text.unicodeScalars.filter { !CharacterSet.controlCharacters.contains($0) || $0 == "\n" || $0 == "\t" })
        return safe.count > 8192 ? String(safe.suffix(8192)) : safe
    }
}

File: native/windows/MightyClaude.Core/CliUpdateService.cs
using System.Text;
using System.Text.Json;
using System.Text.RegularExpressions;

namespace MightyClaude.Core;

/// One recognised installation of a provider CLI.
/// Method is native, winget, npm, unknown or missing.
public sealed record CliUpdateInstallation(
    string Provider,
    string? Version = null,
    string Method = "unknown",
    string? ExecutablePath = null,
    bool CanUpdate = false,
    string Detail = "");

/// The outcome of one update request.
/// Status is updated, current, skipped, failed, cancelled or busy.
/// Output is bounded diagnostic text, shown only on an explicit details action.
public sealed record CliUpdateResult(
    string Provider,
    string Status,
    string? BeforeVersion = null,
    string? AfterVersion = null,
    string Method = "unknown",
    string Detail = "",
    string Output = "");

/// The single invocation chosen for an installation. Never a shell command line:
/// an executable path plus an argument list, handed to ICliRunner as-is.
internal sealed record CliUpdateInvocation(string Executable, IReadOnlyList<string> Arguments, IReadOnlyDictionary<string, string>? Environment = null);

/// Updates only an already installed, recognised CLI through that CLI's own
/// installer. It never installs anything new, never asks for elevation, never
/// writes a configuration file and never upgrades an unrelated package.
/// Every process goes through ICliRunner, so a test injects a fake runner and
/// no real claude, codex, gemini, npm or winget process is started.
public sealed partial class CliUpdateService
{
    private readonly ICliRunner runner;
    private readonly IReadOnlyDictionary<string, string> environment;
    private readonly string home;
    private readonly TimeSpan metadataTimeout;
    private readonly TimeSpan updateTimeout;

    private readonly object gate = new();
    private CancellationTokenSource? active;
    private Task<CliUpdateResult>? activeTask;
    private bool closing;

    // 1 MiB captured by the runner, 8 KiB of it kept for the details action.
    public const int DiagnosticCharacterCap = 8192;
    private const int VersionTextCap = 160;

    public CliUpdateService(
        ICliRunner runner,
        IReadOnlyDictionary<string, string>? environment = null,
        string? homeDirectory = null,
        TimeSpan? metadataTimeout = null,
        TimeSpan? updateTimeout = null)
    {
        this.runner = runner;
        this.environment = environment ?? Runtime();
        home = homeDirectory ?? Environment.GetFolderPath(Environment.SpecialFolder.UserProfile);
        this.metadataTimeout = metadataTimeout ?? TimeSpan.FromSeconds(10);
        this.updateTimeout = updateTimeout ?? TimeSpan.FromMinutes(5);
    }

    private static Dictionary<string, string> Runtime()
    {
        var values = new Dictionary<string, string>(StringComparer.OrdinalIgnoreCase);
        foreach (System.Collections.DictionaryEntry entry in Environment.GetEnvironmentVariables())
            if (entry.Key is string key && entry.Value is string value) values[key] = value;
        return values;
    }

    /// The app checks at start only when the saved switch is a JSON true.
    public static bool ShouldRunAtStartup(AppSnapshot snapshot) => snapshot.AutoUpdateCLIs == true;

    /// macOS ProviderOptions.label — the name the result row shows.
    public static string ProviderLabel(string provider) => provider switch
    {
        "claude" => "Claude",
        "codex" => "Codex",
        "gemini" => "Gemini",
        _ => "Claude",
    };

    /// The row WinUI shows, so no result copy is assembled outside Core.
    public static string ResultRow(CliUpdateResult result) => CliUpdateStrings.ResultRowTemplate
        .Replace("{provider}", ProviderLabel(result.Provider))
        .Replace("{status}", CliUpdateStrings.StatusLabel(result.Status));

    /// The version line, or null when nothing changed.
    public static string? VersionChange(CliUpdateResult result) =>
        result.BeforeVersion is { Length: > 0 } before && result.AfterVersion is { Length: > 0 } after && before != after
            ? CliUpdateStrings.VersionChangeTemplate.Replace("{before}", before).Replace("{after}", after)
            : null;

    public async Task<CliUpdateInstallation> InspectAsync(string provider, CancellationToken cancellation = default)
    {
        if (Closing()) return new CliUpdateInstallation(provider, Detail: CliUpdateStrings.DetailClosing);
        try { return (await PlanAsync(provider, cancellation)).Installation; }
        catch (OperationCanceledException) { return new CliUpdateInstallation(provider, Detail: CliUpdateStrings.DetailInspectCancelled); }
        catch (Exception) { return new CliUpdateInstallation(provider, Detail: CliUpdateStrings.DetailInspectFailed); }
    }

    /// One update at a time. A second request while one runs reports busy.
    public async Task<CliUpdateResult> UpdateAsync(string provider, CancellationToken cancellation = default)
    {
        CancellationTokenSource source;
        Task<CliUpdateResult> task;
        lock (gate)
        {
            if (closing) return new CliUpdateResult(provider, "cancelled", Detail: CliUpdateStrings.DetailClosing);
            if (active is not null) return new CliUpdateResult(provider, "busy", Detail: CliUpdateStrings.DetailBusy);
            if (cancellation.IsCancellationRequested) return new CliUpdateResult(provider, "cancelled", Detail: CliUpdateStrings.DetailCancelled);
            source = CancellationTokenSource.CreateLinkedTokenSource(cancellation);
            active = source;
            task = PerformAsync(provider, source.Token);
            activeTask = task;
        }
        try { return await task; }
        finally
        {
            lock (gate) { if (ReferenceEquals(active, source)) { active = null; activeTask = null; } }
            source.Dispose();
        }
    }

    /// The start-up and "업데이트 하기" pass: one provider after another, never
    /// two installers at once.
    public async Task<IReadOnlyList<CliUpdateResult>> UpdateAllAsync(CancellationToken cancellation = default)
    {
        var results = new List<CliUpdateResult>();
        foreach (var provider in Wire.Providers) results.Add(await UpdateAsync(provider, cancellation));
        return results;
    }

    /// Stops the running installer together with its whole process group and
    /// waits for the result, so the section never shows a half-finished run.
    public async Task CancelAsync()
    {
        Task<CliUpdateResult>? task;
        lock (gate) { active?.Cancel(); task = activeTask; }
        if (task is not null) { try { await task; } catch (Exception) { } }
    }

    /// Closing the app: refuse new work, then cancel cleanly.
    public async Task ShutdownAsync()
    {
        lock (gate) closing = true;
        await CancelAsync();
    }

    private bool Closing() { lock (gate) return closing; }

    private async Task<CliUpdateResult> PerformAsync(string provider, CancellationToken token)
    {
        var installation = new CliUpdateInstallation(provider);
        try
        {
            token.ThrowIfCancellationRequested();
            var (planned, invocation) = await PlanAsync(provider, token);
            installation = planned;
            token.ThrowIfCancellationRequested();
            if (invocation is null)
                return new CliUpdateResult(provider, "skipped", installation.Version, Method: installation.Method, Detail: installation.Detail);

            var run = await runner.RunAsync(invocation.Executable, invocation.Arguments, updateTimeout, token, invocation.Environment);
            token.ThrowIfCancellationRequested();
            var output = Bounded(run);
            if (run.ExitCode != 0)
                return new CliUpdateResult(provider, "failed", installation.Version, Method: installation.Method,
                    Detail: CliUpdateStrings.DetailFailedExitTemplate.Replace("{code}", run.ExitCode.ToString()), Output: output);

            var after = await InstalledCommandAsync(provider, token);
            token.ThrowIfCancellationRequested();
            if (after?.Version is not { Length: > 0 } version)
                return new CliUpdateResult(provider, "failed", installation.Version, Method: installation.Method,
                    Detail: CliUpdateStrings.DetailVersionRecheckFailed, Output: output);

            var changed = installation.Version != version;
            return new CliUpdateResult(provider, changed ? "updated" : "current", installation.Version, version, installation.Method,
                changed ? CliUpdateStrings.DetailUpdated : CliUpdateStrings.DetailUnchanged, output);
        }
        catch (OperationCanceledException)
        {
            return new CliUpdateResult(provider, "cancelled", installation.Version, Method: installation.Method, Detail: CliUpdateStrings.DetailCancelled);
        }
        catch (Exception error)
        {
            var text = error.Message;
            return new CliUpdateResult(provider, "failed", installation.Version, Method: installation.Method,
                Detail: text.Length > 600 ? text[..600] : text);
        }
    }

    private async Task<(CliUpdateInstallation Installation, CliUpdateInvocation? Invocation)> PlanAsync(string provider, CancellationToken token)
    {
        if (!Wire.Providers.Contains(provider))
            return (new CliUpdateInstallation(provider, Detail: CliUpdateStrings.DetailUnsupportedProvider), null);

        var command = await InstalledCommandAsync(provider, token);
        if (command is null)
        {
            var present = Candidates(provider).Any(File.Exists);
            return (new CliUpdateInstallation(provider, Method: present ? "unknown" : "missing",
                Detail: present ? CliUpdateStrings.DetailVersionUnknown : CliUpdateStrings.DetailMissing), null);
        }

        var resolved = Resolve(command.Path);
        var installation = new CliUpdateInstallation(provider, command.Version, "unknown", command.Path, false, CliUpdateStrings.DetailUnknownMethod);

        if (WingetPackage(resolved) is { } package)
        {
            installation = installation with { Method = "winget" };
            if (Candidates("winget").FirstOrDefault(File.Exists) is not { } winget)
                return (installation with { Detail = CliUpdateStrings.DetailWingetRuntimeMissing }, null);
            // Exactly the one package this installation belongs to: --id with
            // --exact, never --all, and no interactive or elevation prompt.
            return (installation with { CanUpdate = true, Detail = CliUpdateStrings.DetailWingetPlan },
                new CliUpdateInvocation(winget, ["upgrade", "--id", package, "--exact", "--silent",
                    "--accept-source-agreements", "--accept-package-agreements", "--disable-interactivity"]));
        }

        if (provider == "claude" && NativeClaude(resolved))
            return (installation with { Method = "native", CanUpdate = true, Detail = CliUpdateStrings.DetailNativeClaude },
                new CliUpdateInvocation(command.Path, ["update"]));

        if (NpmPackage(resolved, provider) is { } npm)
        {
            installation = installation with { Method = "npm" };
            if (!Release().IsMatch(npm.Version))
                return (installation with { Detail = CliUpdateStrings.DetailNpmPrerelease }, null);
            if (FindNpmRuntime(npm.Prefix) is not { } runtime)
                return (installation with { Detail = CliUpdateStrings.DetailNpmRuntimeMissing }, null);
            // The official package only, in the prefix it is already installed in.
            return (installation with { CanUpdate = true, Detail = CliUpdateStrings.DetailNpmPlan },
                new CliUpdateInvocation(runtime.Node, [runtime.Script, "install", "--global", "--prefix", npm.Prefix,
                    npm.Name + "@latest", "--no-audit", "--no-fund"]));
        }

        return (installation, null);
    }

    private sealed record InstalledCommand(string Path, string Version);

    private async Task<InstalledCommand?> InstalledCommandAsync(string provider, CancellationToken token)
    {
        if (!Wire.Providers.Contains(provider)) return null;
        foreach (var path in Candidates(provider))
        {
            token.ThrowIfCancellationRequested();
            if (!File.Exists(path)) continue;
            CliRunResult run;
            try { run = await runner.RunAsync(path, ["--version"], metadataTimeout, token); }
            catch (OperationCanceledException) { throw; }
            catch (Exception) { continue; }
            if (run.ExitCode != 0 || run.TimedOut) continue;
            var text = run.Output.Trim();
            if (text.Length == 0) continue;
            return new InstalledCommand(path, text.Length > VersionTextCap ? text[..VersionTextCap] : text);
        }
        return null;
    }

    // PATH entries, at most 64, each joined with the name and the Windows
    // executable extensions. The bare name is kept last so a fixture without an
    // extension is still found.
    private IEnumerable<string> Candidates(string name)
    {
        var seen = new HashSet<string>(StringComparer.OrdinalIgnoreCase);
        environment.TryGetValue("PATH", out var path);
        foreach (var entry in (path ?? "").Split(System.IO.Path.PathSeparator).Take(64))
        {
            if (entry.Length == 0 || entry.Contains('\0') || !System.IO.Path.IsPathRooted(entry) || !seen.Add(entry)) continue;
            foreach (var extension in new[] { ".cmd", ".exe", ".bat", "" })
                yield return System.IO.Path.Combine(entry, name + extension);
        }
    }

    // winget installs a package under ...\WinGet\Packages\<PackageIdentifier>_<hash>\
    // and puts a link in ...\WinGet\Links. Reading the identifier out of the
    // install location means the upgrade names exactly the package that is
    // already installed instead of a guessed identifier.
    private static string? WingetPackage(string executable)
    {
        var parts = executable.Split(['\\', '/'], StringSplitOptions.RemoveEmptyEntries);
        for (var i = 1; i + 1 < parts.Length; i++)
        {
            if (!parts[i].Equals("Packages", StringComparison.OrdinalIgnoreCase)) continue;
            if (!parts[i - 1].Equals("WinGet", StringComparison.OrdinalIgnoreCase)) continue;
            var folder = parts[i + 1];
            var cut = folder.LastIndexOf('_');
            var id = cut > 0 ? folder[..cut] : folder;
            if (Identifier().IsMatch(id)) return id;
        }
        return null;
    }

    // The native installer keeps its versions under <home>\.local\share\claude\versions
    // (XDG_DATA_HOME when it is set), either as the executable itself or one
    // directory per version, which is the Windows layout.
    private bool NativeClaude(string executable)
    {
        var roots = new List<string> { System.IO.Path.Combine(home, ".local", "share", "claude", "versions") };
        if (environment.TryGetValue("XDG_DATA_HOME", out var data) && data.Length > 0 && System.IO.Path.IsPathRooted(data) && !data.Contains('\0'))
            roots.Add(System.IO.Path.Combine(data, "claude", "versions"));
        var parent = System.IO.Path.GetDirectoryName(executable);
        var grandparent = parent is null ? null : System.IO.Path.GetDirectoryName(parent);
        return roots.Any(root => Same(parent, root) || Same(grandparent, root));
    }

    private static bool Same(string? left, string? right) =>
        left is not null && right is not null &&
        System.IO.Path.TrimEndingDirectorySeparator(Resolve(left)).Equals(System.IO.Path.TrimEndingDirectorySeparator(Resolve(right)), StringComparison.OrdinalIgnoreCase);

    private sealed record NpmInstallation(string Prefix, string Name, string Version);

    // npm --global on Windows puts its shims straight in the prefix folder and
    // the package under <prefix>\node_modules\<name>, so the executable's own
    // folder is the prefix. The package.json must name the official package and
    // declare a bin entry for this CLI, or the install is not recognised.
    private static NpmInstallation? NpmPackage(string executable, string provider)
    {
        var name = provider switch
        {
            "claude" => "@anthropic-ai/claude-code",
            "codex" => "@openai/codex",
            _ => "@google/gemini-cli",
        };
        var prefix = System.IO.Path.GetDirectoryName(executable);
        if (prefix is null) return null;
        var roots = new List<string> { System.IO.Path.Combine(prefix, "node_modules") };
        // A Unix-shaped global prefix keeps the shim in <prefix>/bin.
        var parent = System.IO.Path.GetDirectoryName(prefix);
        if (parent is not null) roots.Add(System.IO.Path.Combine(parent, "lib", "node_modules"));
        foreach (var root in roots)
        {
            var package = System.IO.Path.Combine(root, name.Replace('/', System.IO.Path.DirectorySeparatorChar));
            if (Metadata(package) is not { } json) continue;
            if (json.Name != name || !json.Bin.Contains(provider)) continue;
            var target = System.IO.Path.GetDirectoryName(System.IO.Path.GetDirectoryName(root));
            return new NpmInstallation(root.EndsWith(System.IO.Path.Combine("lib", "node_modules"), StringComparison.OrdinalIgnoreCase) && target is not null
                ? target! : prefix, name, json.Version);
        }
        return null;
    }

    private sealed record PackageJson(string Name, string Version, IReadOnlyCollection<string> Bin);

    private static PackageJson? Metadata(string packageRoot)
    {
        var file = System.IO.Path.Combine(packageRoot, "package.json");
        try
        {
            var info = new FileInfo(file);
            if (!info.Exists || info.Length > 262_144) return null;
            using var json = JsonDocument.Parse(File.ReadAllBytes(file));
            var root = json.RootElement;
            if (root.ValueKind != JsonValueKind.Object) return null;
            var name = root.TryGetProperty("name", out var n) && n.ValueKind == JsonValueKind.String ? n.GetString()! : "";
            var version = root.TryGetProperty("version", out var v) && v.ValueKind == JsonValueKind.String ? v.GetString()! : "";
            var bin = new List<string>();
            if (root.TryGetProperty("bin", out var b))
            {
                if (b.ValueKind == JsonValueKind.Object) bin.AddRange(b.EnumerateObject().Select(p => p.Name));
                else if (b.ValueKind == JsonValueKind.String && name.Length > 0) bin.Add(name[(name.LastIndexOf('/') + 1)..]);
            }
            return new PackageJson(name, version, bin);
        }
        catch (Exception) { return null; }
    }

    private sealed record NpmRuntime(string Node, string Script);

    // npm itself is started through Node with npm's own cli script, never
    // through a .cmd shim, so no shell command line is ever built.
    private NpmRuntime? FindNpmRuntime(string prefix)
    {
        var scripts = new List<string>
        {
            System.IO.Path.Combine(prefix, "node_modules", "npm", "bin", "npm-cli.js"),
            System.IO.Path.Combine(prefix, "lib", "node_modules", "npm", "bin", "npm-cli.js"),
        };
        var script = scripts.FirstOrDefault(File.Exists);
        if (script is null) return null;
        var node = Candidates("node").FirstOrDefault(File.Exists);
        return node is null ? null : new NpmRuntime(node, script);
    }

    private static string Resolve(string path)
    {
        try
        {
            var full = System.IO.Path.GetFullPath(path);
            var target = File.ResolveLinkTarget(full, true) ?? (Directory.Exists(full) ? Directory.ResolveLinkTarget(full, true) : null);
            return target is null ? full : System.IO.Path.GetFullPath(target.FullName);
        }
        catch (Exception) { return path; }
    }

    // stdout then stderr, control characters dropped, the last 8 KiB kept.
    // Shown only when the user asks for details and never saved.
    private static string Bounded(CliRunResult run)
    {
        var text = run.Output + (run.ErrorOutput.Length == 0 ? "" : "\n" + run.ErrorOutput);
        var builder = new StringBuilder(text.Length);
        foreach (var c in text) if (!char.IsControl(c) || c is '\n' or '\t') builder.Append(c);
        var safe = builder.ToString();
        return safe.Length > DiagnosticCharacterCap ? safe[^DiagnosticCharacterCap..] : safe;
    }

    [GeneratedRegex(@"\A[0-9]+\.[0-9]+\.[0-9]+(?:\+[0-9A-Za-z.-]+)?\z")]
    private static partial Regex Release();

    [GeneratedRegex(@"\A[A-Za-z0-9][A-Za-z0-9.+-]{0,127}\z")]
    private static partial Regex Identifier();
}
